### PR TITLE
  Add WebSocket transport implementation for real-time communication

### PR DESCRIPTION
**Pull request by alimoradi296**
Implements comprehensive WebSocket transport following UTCP architecture with reviewer feedback addressed.

  ## Core Features
  - Real-time bidirectional communication via WebSocket protocol
  - Tool discovery through WebSocket handshake using UTCP messages
  - Streaming tool execution with proper error handling
  - Connection management with keep-alive and reconnection support

  ## Architecture Compliance
  - Dependency injection pattern with constructor injection
  - Implements ClientTransportInterface contract
  - Composition over inheritance design
  - Clear separation of data and business logic
  - Thread-safe and scalable implementation

  ## Authentication & Security
  - Full authentication support (API Key, Basic Auth, OAuth2)
  - Security enforcement (WSS required, localhost exception)
  - Custom headers and protocol specification support

  ## Reviewer Feedback Addressed
  - **Individual tool providers**: Each tool now gets its own WebSocketProvider instance instead of sharing the manual provider
  - **Flexible message format**: Added `message_format` field for custom WebSocket message formatting to support existing services
  without modification
  - **New branch**: Created `feature/websockets` branch as requested by @h3xxit

  ## Testing & Quality
  - Unit tests covering all functionality (80%+ coverage)
  - Mock WebSocket server for development/testing
  - Integration with existing UTCP test patterns
  - Comprehensive error handling and edge cases

  ## Protocol Implementation
  - Discovery: `{"type": "discover", "request_id": "id"}`
  - Tool calls: `{"type": "call_tool", "tool_name": "name", "arguments": {...}}`
  - Responses: `{"type": "tool_response|tool_error", "result": {...}}`
  - Custom formats supported via `message_format` template

  ## Documentation
  - Complete example with interactive client/server demo
  - Updated README removing "work in progress" status
  - Protocol specification and usage examples

  Addresses the "No wrapper tax" principle by enabling direct WebSocket communication without requiring changes to existing
  WebSocket services. Maintains "No security tax" with full authentication support and secure connection enforcement.

  Replaces PR #33 with the requested `feature/websockets` branch.
  updates:
    - ✅ Resolved merge conflicts with upstream/dev
  - ✅ Aligned WebSocket implementation with official UDP pattern
  - ✅ Maintained backward compatibility while adding new features
  - ✅ All tests passing (WebSocket, TCP, UDP)

### CODE REVIEW

**Review by h3xxit — CHANGES_REQUESTED** [review]
Update to 1.0.0

**Comment by alimoradi296** [comment]
@h3xxit I've updated the WebSocket implementation to UTCP 1.0.0! 🎉

  ## Changes Made

  ✅ **Plugin Architecture**
  - Migrated to new plugin structure following HTTP/CLI/MCP patterns
  - Implements `CommunicationProtocol` interface with all required methods
  - Entry point registration: `websocket = "utcp_websocket:register"`

  ✅ **1.0.0 Compliance**
  - `Provider` → `CallTemplate` pattern (`WebSocketCallTemplate`)
  - `register_tool_provider()` → `register_manual()` returns `RegisterManualResult`
  - Discovery returns complete `UtcpManual` format
  - All imports updated to new module structure

  ✅ **Testing & Documentation**
  - 8 comprehensive tests (100% passing)
  - Complete README with configuration examples
  - Added to main README protocol plugins table as ✅ Stable

  ✅ **Features Maintained**
  - Real-time bidirectional WebSocket communication
  - Authentication (API Key, Basic Auth, OAuth2)
  - Flexible message formats (JSON + text templates with `UTCP_ARG_` placeholders)
  - Connection pooling, keep-alive, and security enforcement

  ## Key Commits

  - `7a31f82` - Migrate WebSocket transport to UTCP 1.0.0 plugin architecture
  - `86389da` - Add WebSocket plugin tests and update main README

  ## Plugin Structure

  plugins/communication_protocols/websocket/
  ├── src/utcp_websocket/
  │   ├── init.py                              # Plugin registration
  │   ├── websocket_call_template.py               # CallTemplate with validation
  │   └── websocket_communication_protocol.py      # CommunicationProtocol implementation
  ├── tests/
  │   ├── init.py
  │   └── test_websocket_call_template.py          # 8 tests, all passing
  ├── pyproject.toml                               # Package config with entry points
  └── README.md                                    # Complete documentation

  Ready for review! Let me know if you need any changes.

**Comment by alimoradi296** [comment]
@h3xxit Thanks for the feedback! I've addressed the flexibility concerns by completely refactoring the WebSocket
  implementation.

  ## Key Changes

  **Maximum Flexibility Approach:**
  - ✅ No enforced request/response format - works with ANY WebSocket endpoint
  - ✅ Default behavior: sends arguments as-is in JSON, returns raw responses
  - ✅ Optional `message` template field for endpoints requiring specific formats
  - ✅ Optional `response_format` field for structured response parsing

  **What Changed:**
  - Removed restrictive fields: `request_data_format`, `request_data_template`, `message_format`
  - Added flexible `message` field (Union[str, Dict]) with `${arg_name}` placeholders
  - Added `response_format` field (Optional["json", "text", "raw"])
  - Updated all tests (9/9 passing)
  - Updated documentation to emphasize flexibility

  **Examples:**

  No template (works with any endpoint):
  ```json
  {"call_template_type": "websocket", "url": "wss://api.example.com/ws"}
  Sends: {"user_id": "123", "action": "getData"}

  Dict template (structured protocols):
  {"message": {"type": "request", "action": "${action}", "params": "${params}"}}

  String template (text-based protocols):
  {"message": "CMD:${command};DEVICE:${device_id};VALUE:${value}"}

  The philosophy is: UTCP adapts to existing endpoints, not the other way around.

  Ready for review!

File: plugins/communication_protocols/websocket/src/utcp_websocket/websocket_call_template.py [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+        timeout: Timeout in seconds for WebSocket operations.
+        headers: Optional static headers to include in WebSocket handshake.
+        header_fields: List of tool argument names to map to WebSocket handshake headers.
+        auth: Optional authentication configuration for WebSocket connection.
+    """
+    call_template_type: Literal["websocket"] = Field(default="websocket")
+    url: str = Field(..., description="WebSocket URL (wss:// or ws://localhost)")
+    message: Optional[Union[str, Dict[str, Any]]] = Field(
+        default=None,
+        description="Message template. Can be a string or dict with ${arg_name} placeholders"
```
Comment (h3xxit) [inline plugins/communication_protocols/websocket/src/utcp_websocket/websocket_call_template.py:71]: The message template has to have args delimited by UTCP_ARG_ and _UTCP_ARG because ${ and } is already reserved for the secret variable replacement. So the client will automatically try and replace ${} with variables from .env and other secrets managers, and throw an error if it doesnt find them.

**Review by h3xxit — CHANGES_REQUESTED** [review]
Very good stuff! I've left one comment on the formatting delimitors. Check that and the cubic suggestions, and once thats done, we can merge it!

**Review by h3xxit — APPROVED** [review]
@alimoradi296 Perfect! Thanks a lot for all the work.

Please let me know your name and any links you want added to the www.utcp.io/about page. Additionally if you want to join the discord community I'll give you a contributor role there too: https://discord.gg/ZpMbQ8jRbD
